### ISSUE TITLE
Incorrect IHasPoints.GetPointNearestX results using SignalPlotXY 

### ISSUE THREAD

**Issue by at2software** [ScottPlot/ScottPlot#809]
ScottPlot version v4.1.8-beta

_Note: this might be related to #548 and maybe this function is just not yet supported for `SignalPlotXY` (then this is not a bug of course, but probably this function should then not be available to `SignalPlotXY` instead of returning false values)._

There seems to be an issue with the results returned by `IHasPoints.GetPointNearestX` with a `SignalPlotXY` when the x values aren't consecutive integers starting at zero. 
Here is an example to highlight the nearest point of a ScatterPlot (adapted from [here](https://github.com/ScottPlot/ScottPlot/blob/b09e6dd902378c5818e64ed8efb81ae408b40e37/src/sandbox/WinFormsFrameworkApp/Form1.cs))

```cs
public partial class MainWindow : Window
{
	private readonly IHasPoints plottable;
	private readonly ScatterPlot HighlightedPoint;
	private int LastHighlightedIndex = -1;

	public MainWindow()
	{
		InitializeComponent();

		int sampleCount = 50;
		double[] xs = new double[sampleCount];
		double[] ys = new double[sampleCount];

		Random rand = new Random(0);
		for (int i = 0; i < 50; i++)
		{
			xs[i] = DateTime.Now.ToOADate() + i*10;
			ys[i] = i > 0 ? ys[i - 1] + rand.NextDouble() - .5 : 0;
		}

		//plottable = plot.Plot.AddScatterPoints(xs, ys);    // <-- this works as expected
		plottable = plot.Plot.AddSignalXY(xs, ys);
			
		HighlightedPoint = plot.Plot.AddPoint(0, 0);
		HighlightedPoint.Color = System.Drawing.Color.Red;
		HighlightedPoint.MarkerSize = 10;
		HighlightedPoint.MarkerShape = MarkerShape.openCircle;
		HighlightedPoint.IsVisible = false;

		plot.Plot.XAxis.DateTimeFormat(true);
	}

	private void plot_MouseMove(object sender, MouseEventArgs e)
	{
		(double mouseCoordX, double mouseCoordY) = plot.GetMouseCoordinates();

		// these results are incorrect when plottable is SignalPlotXY:
		(double pointCoordX, double pointCoordY, int pointIndex) = plottable.GetPointNearestX(mouseCoordX);

		// highlight the point of interest
		HighlightedPoint.Xs[0] = pointCoordX;
		HighlightedPoint.Ys[0] = pointCoordY;
		HighlightedPoint.IsVisible = true;

		tbPointCoord.Text = $"{pointCoordX:N1}, {pointCoordY:N1}";
		tbIndex.Text = pointIndex.ToString();

		if (LastHighlightedIndex != pointIndex)
		{
			LastHighlightedIndex = pointIndex;
			plot.Render();
		}
	}
}
```

With the ScatterPlot this works as expected. Changing the `IHasPoints` plottable to `SignalPlotXY`, `GetPointNearestX` now gives wrong results, the index always relates to x=0. This is especially striking when using `DateTimeFormat(true)` as here we provide `DateTime.ToOADate()` values as xs values, which e.g. are in the range of 44000 and above. So here, the determined index by `GetPointNearestX` is always the max index of the array (49 in the example case).

I guess this is because here https://github.com/ScottPlot/ScottPlot/blob/92ec0558a76795a2fbfc4a54258399522092b074/src/ScottPlot/Plottable/SignalPlotBase.cs#L689 it is assumed that the x parameter is directly correlated to the index, which is true for `SignalPlot` but not for `SignalPlotXY`. Probably `GetPointNearestX` should be made virtual in `SignalPlotBase` and `SignalPlotXY` would need to override it with an implementation to determine the index based on the parameter value similar to here https://github.com/ScottPlot/ScottPlot/blob/92ec0558a76795a2fbfc4a54258399522092b074/src/ScottPlot/Plottable/ScatterPlot.cs#L292-L309

**Comment by StendProg** [comment]
Hi @at2software,
As you rightly noted, `SignalXY` does not support. GetPointNearestX It uses the implementation of the parent which is `PlottableSignal`. And of course it does not work correctly because `SignalXY` uses a completely different approach to X values.
There is possible to exclude support for the `IHasPoints` interface from `PlottableSignalXY`.
But it may be possible to implement it's support, I'll see what can be done the other day.

**Comment by at2software** [comment]
Thanks for the clarification @StendProg.

Here's a proposed solution for this issue: For `SignalPlotXY`, only ascending x values are allowed, so the `xs[]` array is automatically sorted. Hence we can apply binary search, as opposed to iterating linearly, which cuts the complexity down to **O(log n)** instead of **O(n)**. We thus can quite efficiently obtain the nearest point, sort of like this:

* make `GetPointNearestX` virtual in `SignalPlotBase`
* override `GetPointNearestX` in `SignalPlotXY` or `SignalPlotXYGeneric`
* implement `GetPointNearestX` sort of like this:

```cs
public override (double x, double y, int index) GetPointNearestX(double x)
{
	int index = Array.BinarySearch(Xs, x);
	if (index < 0)
	{
		index = ~index - 1;
	}
	index = Math.Min(Math.Max(index, 0), Xs.Length - 2);

	double distLeft  = Math.Abs(Xs[index] - x);
	double distRight = Math.Abs(Xs[index + 1] - x);
	if (distRight < distLeft)
	{
		index++;
	}
	return (Xs[index], Ys[index], index);
}
```

A quick test looks promising, see this `SignalPlotXY` with 5 million values and above strategy:
![2021-02-19_10_11_03_3g20 mp4](https://user-images.githubusercontent.com/8616861/108483538-23bcb880-729b-11eb-9d37-842cf528d67b.gif)

Here's a performance comparison between the above mentioned modified binary search strategy versus the iterative version (linear measurement of each distance between `MinRenderIndex` and `MaxRenderIndex`):

![2021-02-19_10_29_01_vt80 mp4](https://user-images.githubusercontent.com/8616861/108485629-6a131700-729d-11eb-9dc9-c215b4d47d94.gif)

**Note** that signal specifics like `OffsetX/Y` and `SamplePeriod` are not respected yet. I'm not yet far enough in the ScottPlot code, but probably you or someone else can make these additions and make this a robust solution.

**Comment by StendProg** [comment]
Hi @at2software,
`SignalXY` already uses binary search to find the indices of the array X. So it is logical to implement `GetPointNearestX` using binary search as well.

It looks like you have worked out this issue in great detail, maybe you will make a pull request?
I have only one suggestion for the given code, there is a suggestion to return the result immediately when it is ready.
For example, when a binary search returns a positive value, i.e. `x == X[index]`, return it immediately.
When the bitwise complement value is less than 0, then also immediately return the 0 element. Same with the right border. It seems to me that this will slightly simplify the logic (and then I'm not sure until try it).

What do you think?

**Comment by at2software** [comment]
Ah yes, I didn't really look quite deeply into the SignalPlotXY implementation (rather just posted our tests and workaround implementation from our project), the usage of BinarySearch is quite similarly already there, correct.

Your proposed early exits for an exact match and first / last points are reasonable of course.

About the PR: I think one should also take `OffsetX`, `OffsetY` and `SamplePeriod` into account (maybe something else as well? `MinRenderIndex` and `MaxRenderIndex`, e.g. must be respected, too, I guess). And as I'm quite unfamiliar with the ScottPlot code (as could be seen 😄), I'm not certain to deliver a really general and robust implementation, I think. Maybe there's not much to add besides adding offsets, dividing by the Period and clamping to Min-/MaxRenderIndex, but honestly I don't really know and probably don't have test scenarios outside our limited application.

**Comment by swharden** [comment]
Following-up on this, 

I think the original issue was solved by #886 which was merged today and I'll release on NuGet tonight.

SignalPlotXY does not properly support OffsetX and OffsetY, so I created an issue to track that #890

I think this is enough so this ticket can be closed, but let me know if any other questions or issues come up. Thanks for both of your input on this topic!
